shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,790 @@
-  var stackedUsers = {};
-  stackedUsers[openerIdentityId] = true;
-  var userStack = [openerIdentityId];
-
-  while (userStack.length > 0) {
-    var recipient = userStack.pop();
-    var edges = edgesByRecipient[recipient];
-    if (edges) {
-      for (var ii = 0; ii < edges.length; ++ii) {
-        var inEdge = edges[ii];
-        var sharer = inEdge.sharer;
-        if (sharer == owner) {
-          return true;
+  const stack = [chooseAll];
+  // Array of Map<hashedClause, { clause: Clause, tokensUsed: [String] }>.
+
+  while (stack.length > 0) {
+    const chosen = stack.pop();
+    let minimal = true;
+    for (const removedHashedClause in chosen) {
+      const newChosen = {};
+      for (const hashedClause in chosen) {
+        if (hashedClause !== removedHashedClause) {
+          newChosen[hashedClause] = chosen[hashedClause];
         }
-        if (!stackedUsers[sharer]) {
-          userStack.push(sharer);
-          stackedUsers[sharer] = true;
+      }
+
+      if (fullCover(newChosen)) {
+        minimal = false;
+        // TODO(perf): keep track of values we've already explored.
+        stack.push(newChosen);
+      }
+    }
+
+    if (minimal) {
+      const resultClause = new Clause();
+      const tokensUsed = {};
+      for (const hashedClause in chosen) {
+        resultClause.conjoin(chosen[hashedClause].clause);
+        for (const tokenId in chosen[hashedClause].tokensUsed) {
+          tokensUsed[tokenId] = true;
         }
       }
+
+      result[resultClause.hash()] = { clause: resultClause, tokensUsed: tokensUsed };
+    }
+  }
+
+  return result;
+}
+
+function proveClauses(db, context, goalClauses) {
+  // `goalClauses` is of the form of a result of `normalize()`, that is:
+  // `Map<ClauseId, {clause: Clause, tokensUsed: [tokenId]}>`.
+  //
+  // This function attempts to determine whether at least one of the input clauses holds.
+  // If a proof is found, returns `{yes: {tokensUsed: [tokenId]}}`.
+  // otherwise, returns {no: {}}
+  //
+  // The `db` parameter is optional. If it is present, the database will be queried as needed.
+  // If left out, the computation will proceed with only the tokens already present in `context`.
+
+  const clauses = {};
+  const clauseStack = []; // clause IDs
+
+  for (const hashedClause in goalClauses) {
+    const goalClause = goalClauses[hashedClause];
+    const tokensUsed = goalClause.tokensUsed;
+    if (goalClause.clause.isEmpty()) {
+      // We have an empty goal. Trivially true.
+      return { yes: { tokensUsed: tokensUsed } };
+    }
+
+    clauses[hashedClause] = { clause: goalClause.clause, tokensUsed: tokensUsed };
+    clauseStack.push(hashedClause);
+  }
+
+  const grainsFullyQueried = {};
+
+  while (clauseStack.length > 0) {
+    const hashedClause = clauseStack.pop();
+    const clause = new Clause();
+    clause.conjoin(clauses[hashedClause].clause);
+    const tokensUsed = clauses[hashedClause].tokensUsed;
+    const goal = clause.popFirstGoal();
```

Comment:
It seems like some memoization of goals is needed here, otherwise you could easily end up checking the same goal over and over as it appears in different clauses.